{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import requests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Querying Signature Commons\n",
    "Signature Commons includes two API that users can interact with: (1) the metadata API (https://maayanlab.cloud/sigcom/metadata-api), and (2) the data API (https://maayanlab.cloud/sigcom/data-api). The metadata API provides fast full-text search and field comparison filtering of the metadata as well as metadata aggregations for statistical summaries. Users can query the following models (1) resources, (2) libraries, (3) signatures, (4) entities. Meanwhile, the data API performs real-time set-, ranked set-, and two-sided ranked set-enrichment analysis. Here we show some of the basic use cases for the two APIs.\n",
    "\n",
    "## Metadata Search\n",
    "Users can query for metadata using JSON formatted loopback queries (https://loopback.io/doc/en/lb3/Querying-data.html). \n",
    "\n",
    "\n",
    "**Method** \t`GET`<br>\n",
    "**URL** \t`/<model>`\n",
    "\n",
    "\n",
    "**PARAMETERS**<br>\n",
    "**filter** \"stringified JSON query\n",
    "\n",
    "**RETURNS** JSON list of all the matches\n",
    "\n",
    "### Full Text Search"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\n",
      "    {\n",
      "        \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "        \"id\": \"10cd4a9a-c1a1-11e8-8c55-0242ac170004\",\n",
      "        \"library\": \"30b4aba6-d3e2-11e8-a3b8-787b8ad942f3\",\n",
      "        \"meta\": {\n",
      "            \"Assay\": \"microarray\",\n",
      "            \"geo_id\": \"GSE28698\",\n",
      "            \"Readout\": \"mRNA\",\n",
      "            \"Organism\": \"Homo sapiens\",\n",
      "            \"Direction\": \"down\",\n",
      "            \"$validator\": \"/@dcic/signature-commons-schema/v4/meta/signature/sigcom.json\",\n",
      "            \"Small_Molecule\": [\n",
      "                {\n",
      "                    \"Name\": \"imatinib\",\n",
      "                    \"Accession\": \"CHEBI:45783\",\n",
      "                    \"DrugBankID\": \"DB00619\"\n",
      "                }\n",
      "            ],\n",
      "            \"Number_of_genes\": 375,\n",
      "            \"Original_String\": \"imatinib DB00619 human GSE28698 sample 2524\",\n",
      "            \"Perturbation_Type\": \"small molecule\"\n",
      "        }\n",
      "    }\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "METADATA_API = \"https://maayanlab.cloud/sigcom/metadata-api\"\n",
    "endpoint = \"/signatures\"\n",
    "term = \"imatinib\"\n",
    "\n",
    "query = {\n",
    "    \"where\": {\n",
    "        \"meta\": {\"fullTextSearch\": term}     \n",
    "    },\n",
    "    \"limit\": 1\n",
    "}\n",
    "res = requests.get(METADATA_API+ endpoint , params={\"filter\": json.dumps(query)})\n",
    "print(json.dumps(res.json(),indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alternatively, users can also perform a POST request using the `/find` endpoint\n",
    "\n",
    "**Method** \t`POST`<br>\n",
    "**URL** \t`/<model>/find`\n",
    "\n",
    "\n",
    "**PARAMETERS**<br>\n",
    "**filter** JSON serialized query\n",
    "\n",
    "**RETURNS** JSON list of all the matches\n",
    "\n",
    "### Filtering by fields"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\n",
      "    {\n",
      "        \"$validator\": \"/dcic/signature-commons-schema/v5/core/library.json\",\n",
      "        \"id\": \"30f1072b-d3e2-11e8-bf7f-787b8ad942f3\",\n",
      "        \"resource\": \"5aa7c5f3-b48f-4f02-8737-3b54b13ef286\",\n",
      "        \"dataset\": \"enrichr_geneset\",\n",
      "        \"dataset_type\": \"geneset_library\",\n",
      "        \"meta\": {\n",
      "            \"DOI\": \"10.1039/c4mb00677a\",\n",
      "            \"Icon\": \"static/images/LINCS_L1000_Ligand_Perturbations_down.ico\",\n",
      "            \"PMID\": \"25609570\",\n",
      "            \"Spec\": \"(?P<Original_String>((?P<UUID>.*):SC=)?(?P<Gene>.+?)-(?P<Cell_Line>.+?))\\\\t\\\\t(?P<Genes>.+)\",\n",
      "            \"Assay\": \"L1000\",\n",
      "            \"Program\": \"LINCS\",\n",
      "            \"Readout\": \"mRNA\",\n",
      "            \"Version\": \"2014\",\n",
      "            \"Filename\": \"LINCS_L1000_Ligand_Perturbations_down\",\n",
      "            \"Organism\": \"Homo sapiens\",\n",
      "            \"Weighted\": \"False\",\n",
      "            \"Direction\": \"up\",\n",
      "            \"$validator\": \"/@dcic/signature-commons-schema/v4/meta/library/sigcom.json\",\n",
      "            \"Access_URL\": \"http://amp.pharm.mssm.edu/Enrichr/geneSetLibrary?mode=text&libraryName=LINCS_L1000_Ligand_Perturbations_down\",\n",
      "            \"Description\": \"L1000 is a gene expression profiling assay based on the direct measurement of a reduced representation of 1000 genes to represent the transcriptome and computational inference of the portion of the transcriptome not explicitly measured. This subset of the data only includes experiments with drugs that target ligands.\",\n",
      "            \"Library_name\": \"LINCS L1000 Ligand Perturbations down\",\n",
      "            \"Original_source\": \"CMAP\",\n",
      "            \"Link_to_resource\": \"https://clue.io/\",\n",
      "            \"Primary_Resource\": \"CMAP\",\n",
      "            \"Perturbation_Type\": \"ligand perturbation\",\n",
      "            \"Library_created_by\": \"MaayanLab\"\n",
      "        }\n",
      "    }\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "METADATA_API = \"https://maayanlab.cloud/sigcom/metadata-api\"\n",
    "endpoint = \"/libraries/find\"\n",
    "term = \"L1000\"\n",
    "\n",
    "query = {\n",
    "    \"filter\":{\n",
    "        \"where\": {\n",
    "            \"meta.Assay\": term     \n",
    "        },\n",
    "        \"limit\": 1\n",
    "    }\n",
    "}\n",
    "res = requests.post(METADATA_API+ endpoint, json=query)\n",
    "print(json.dumps(res.json(),indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aggregating metadata\n",
    "\n",
    "### Counting results\n",
    "\n",
    "**Method** \t`GET`<br>\n",
    "**URL** \t`/<model>/count`\n",
    "\n",
    "\n",
    "**PARAMETERS**<br>\n",
    "**filter** JSON serialized query\n",
    "\n",
    "**RETURNS** The number of matching results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"count\": 183\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "METADATA_API = \"https://maayanlab.cloud/sigcom/metadata-api\"\n",
    "endpoint = \"/signatures/count\"\n",
    "term = \"STAT3\"\n",
    "\n",
    "query = {\n",
    "    \"meta\": {\"fullTextSearch\": term}     \n",
    "}\n",
    "res = requests.get(METADATA_API+ endpoint,  params={\"where\": json.dumps(query)})\n",
    "print(json.dumps(res.json(),indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Counting top terms\n",
    "\n",
    "**Method** \t`GET`<br>\n",
    "**URL** \t`/<model>/value_count`\n",
    "\n",
    "\n",
    "**PARAMETERS**<br>\n",
    "**filter** JSON serialized query\n",
    "\n",
    "**RETURNS** JSON list of the value counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"meta.Assay\": {\n",
      "        \"L1000\": 3263,\n",
      "        \"microarray\": 21,\n",
      "        \"RNA-seq\": 2,\n",
      "        \"ChIP-seq\": 2\n",
      "    }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "METADATA_API = \"https://maayanlab.cloud/sigcom/metadata-api\"\n",
    "endpoint = \"/signatures/value_count\"\n",
    "term = \"MCF10A\"\n",
    "\n",
    "query = {\n",
    "    \"fields\": [\"meta.Assay\"],\n",
    "    \"where\": {\n",
    "        \"meta\": {\"fullTextSearch\": term}     \n",
    "    }\n",
    "}\n",
    "res = requests.get(METADATA_API+ endpoint,  params={\"filter\": json.dumps(query)})\n",
    "print(json.dumps(res.json(),indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Counting top terms\n",
    "\n",
    "**Method** \t`GET`<br>\n",
    "**URL** \t`/<model>/key_count`\n",
    "\n",
    "\n",
    "**PARAMETERS**<br>\n",
    "**filter** JSON serialized query\n",
    "\n",
    "**RETURNS** JSON list of the value counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"meta.Assay\": 9947,\n",
      "    \"meta.Organism\": 9947,\n",
      "    \"meta.Readout\": 9947,\n",
      "    \"meta.Number_of_genes\": 9947,\n",
      "    \"id\": 9947,\n",
      "    \"meta.$validator\": 9947,\n",
      "    \"library\": 9947,\n",
      "    \"meta.Original_String\": 9947,\n",
      "    \"meta.Perturbation_Type\": 9947,\n",
      "    \"meta.Direction\": 7138,\n",
      "    \"meta.Tissue.Name\": 6917,\n",
      "    \"meta.Tissue.Accession\": 6901,\n",
      "    \"meta.Age\": 5893,\n",
      "    \"meta.Sex\": 5844,\n",
      "    \"meta.Program\": 5836,\n",
      "    \"meta.Patient_ID\": 5836,\n",
      "    \"meta.Gene.Name\": 3362,\n",
      "    \"meta.Gene.Accession\": 2950,\n",
      "    \"meta.Gene_Family\": 2574,\n",
      "    \"meta.geo_id\": 1302,\n",
      "    \"meta.Gene.Gene_Perturbation\": 746,\n",
      "    \"meta.Cell_Line.Name\": 462,\n",
      "    \"meta.Cell_Line.Accession\": 452,\n",
      "    \"meta.Small_Molecule.Name\": 396,\n",
      "    \"meta.Time_Point\": 344,\n",
      "    \"meta.Small_Molecule.Accession\": 316,\n",
      "    \"meta.Small_Molecule.Concentration\": 208,\n",
      "    \"meta.Concentration\": 200,\n",
      "    \"meta.Disease.Name\": 150,\n",
      "    \"meta.Disease.Accession\": 148,\n",
      "    \"meta.Mouse_Model\": 36,\n",
      "    \"meta.Disease.Remarks\": 16,\n",
      "    \"meta.Other.Name\": 10,\n",
      "    \"meta.Other.Accession\": 6,\n",
      "    \"meta.Tissue.Remarks\": 4,\n",
      "    \"meta.Virus.Name\": 2,\n",
      "    \"meta.Other.Remarks\": 2,\n",
      "    \"meta.Virus.Accession\": 2,\n",
      "    \"meta.Library\": 1\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "METADATA_API = \"https://maayanlab.cloud/sigcom/metadata-api\"\n",
    "endpoint = \"/signatures/key_count\"\n",
    "term = \"RNA-seq\"\n",
    "\n",
    "query = {\n",
    "    \"where\": {\n",
    "        \"meta\": {\"fullTextSearch\": term}     \n",
    "    }\n",
    "}\n",
    "res = requests.get(METADATA_API+ endpoint,  params={\"filter\": json.dumps(query)})\n",
    "print(json.dumps(res.json(),indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Signature Search\n",
    "Performing signature search requires a few steps that requires the users to query both the metadata and data API. The following demonstrates how to perform signature search using (1) unranked gene sets, and (2) up and down gene sets.\n",
    "\n",
    "### Getting the list of datasets\n",
    "Users must specify the dataset they want to perform enrichment analysis on. To figure out the list of available datasets, one can use the `/listdata` endpoint of the data api.\n",
    "\n",
    "**Method** \t`GET`<br>\n",
    "**URL** \t`/listdata`\n",
    "\n",
    "\n",
    "**RETURNS** A JSON list of datasets stating the identifier for the dataset and the datatype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"repositories\": [\n",
      "        {\n",
      "            \"datatype\": \"geneset_library\",\n",
      "            \"uuid\": \"enrichr_geneset\"\n",
      "        },\n",
      "        {\n",
      "            \"datatype\": \"geneset_library\",\n",
      "            \"uuid\": \"creeds_geneset\"\n",
      "        },\n",
      "        {\n",
      "            \"datatype\": \"rank_matrix\",\n",
      "            \"uuid\": \"lincs_clue\"\n",
      "        }\n",
      "    ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "DATA_API = \"https://maayanlab.cloud/sigcom/data-api/api/v1\"\n",
    "endpoint = \"/listdata\"\n",
    "\n",
    "res = requests.get(DATA_API+ endpoint)\n",
    "print(json.dumps(res.json(),indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Converting gene names to UUIDs\n",
    "Genes should be converted to UUIDs before sending them to the data api for signature search. To do this, we need to perform metadata search on the gene name field"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['c998bbe8-ed1c-11e8-a7d6-787b8ad942f3', 'c98d5e1a-ed1c-11e8-a48a-787b8ad942f3', 'c99367c6-ed1c-11e8-a4b9-787b8ad942f3']\n"
     ]
    }
   ],
   "source": [
    "METADATA_API = \"https://maayanlab.cloud/sigcom/metadata-api\"\n",
    "endpoint = \"/entities/find\"\n",
    "genes = ['UTP14A','S100A6','SCAND1']\n",
    "\n",
    "query = {\n",
    "    \"filter\":{\n",
    "        \"where\": {\n",
    "            \"meta.Name\": {\n",
    "                \"inq\": genes\n",
    "            }     \n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "res = requests.post(METADATA_API+ endpoint, json=query)\n",
    "gene_ids = [i[\"id\"] for i in res.json()]\n",
    "print(gene_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The way a user perform enrichment analysis is determined by the type of input they are using to query sigcom and the datatype of the datasets.\n",
    "\n",
    "### Unranked Gene Set\n",
    "#### Datatype: geneset_library\n",
    "**Method** \t`POST`<br>\n",
    "**URL** \t`/enrich/overlap`\n",
    "\n",
    "\n",
    "**Query**<br>\n",
    "```\n",
    "{\n",
    "  \"database\": \"string\",\n",
    "  \"entities\": [\n",
    "    \"string\"\n",
    "  ],\n",
    "  \"offset\": 0,\n",
    "  \"limit\": 0\n",
    "}\n",
    "```\n",
    "\n",
    "**Returns** The analysis results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\n",
      "    {\n",
      "        \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "        \"id\": \"34bf2434-c1a1-11e8-aaf5-0242ac170004\",\n",
      "        \"library\": \"30e1e0a8-d3e2-11e8-8014-787b8ad942f3\",\n",
      "        \"meta\": {\n",
      "            \"Assay\": \"text mining\",\n",
      "            \"Tissue\": [\n",
      "                {\n",
      "                    \"Name\": \"islet of Langerhans\",\n",
      "                    \"Accession\": \"UBERON:0000006\"\n",
      "                }\n",
      "            ],\n",
      "            \"Readout\": \"mRNA\",\n",
      "            \"Organism\": \"Homo sapiens\",\n",
      "            \"$validator\": \"/@dcic/signature-commons-schema/v4/meta/signature/sigcom.json\",\n",
      "            \"Number_of_genes\": 4972,\n",
      "            \"Original_String\": \"Pancreatic islet\",\n",
      "            \"Perturbation_Type\": \"shared location\"\n",
      "        }\n",
      "    },\n",
      "    {\n",
      "        \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "        \"id\": \"34b90780-c1a1-11e8-92dc-0242ac170004\",\n",
      "        \"library\": \"30e1e0a8-d3e2-11e8-8014-787b8ad942f3\",\n",
      "        \"meta\": {\n",
      "            \"Assay\": \"text mining\",\n",
      "            \"Tissue\": [\n",
      "                {\n",
      "                    \"Name\": \"intestine\",\n",
      "                    \"Accession\": \"UBERON:0000160\"\n",
      "                }\n",
      "            ],\n",
      "            \"Readout\": \"mRNA\",\n",
      "            \"Organism\": \"Homo sapiens\",\n",
      "            \"$validator\": \"/@dcic/signature-commons-schema/v4/meta/signature/sigcom.json\",\n",
      "            \"Number_of_genes\": 5377,\n",
      "            \"Original_String\": \"Intestine\",\n",
      "            \"Perturbation_Type\": \"shared location\"\n",
      "        }\n",
      "    },\n",
      "    {\n",
      "        \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "        \"id\": \"34c6b4c2-c1a1-11e8-bf22-0242ac170004\",\n",
      "        \"library\": \"30e1e0a8-d3e2-11e8-8014-787b8ad942f3\",\n",
      "        \"meta\": {\n",
      "            \"Assay\": \"text mining\",\n",
      "            \"Tissue\": [\n",
      "                {\n",
      "                    \"Name\": \"thyroid gland\",\n",
      "                    \"Accession\": \"UBERON:0002046\"\n",
      "                }\n",
      "            ],\n",
      "            \"Readout\": \"mRNA\",\n",
      "            \"Organism\": \"Homo sapiens\",\n",
      "            \"$validator\": \"/@dcic/signature-commons-schema/v4/meta/signature/sigcom.json\",\n",
      "            \"Number_of_genes\": 13652,\n",
      "            \"Original_String\": \"Thyroid gland\",\n",
      "            \"Perturbation_Type\": \"shared location\"\n",
      "        }\n",
      "    },\n",
      "    {\n",
      "        \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "        \"id\": \"34af102c-c1a1-11e8-8065-0242ac170004\",\n",
      "        \"library\": \"30e1e0a8-d3e2-11e8-8014-787b8ad942f3\",\n",
      "        \"meta\": {\n",
      "            \"Assay\": \"text mining\",\n",
      "            \"Tissue\": [\n",
      "                {\n",
      "                    \"Name\": \"adipose tissue\",\n",
      "                    \"Accession\": \"UBERON:0001013\"\n",
      "                }\n",
      "            ],\n",
      "            \"Readout\": \"mRNA\",\n",
      "            \"Organism\": \"Homo sapiens\",\n",
      "            \"$validator\": \"/@dcic/signature-commons-schema/v4/meta/signature/sigcom.json\",\n",
      "            \"Number_of_genes\": 12174,\n",
      "            \"Original_String\": \"Adipose tissue\",\n",
      "            \"Perturbation_Type\": \"shared location\"\n",
      "        }\n",
      "    },\n",
      "    {\n",
      "        \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "        \"id\": \"34c2b63a-c1a1-11e8-bda2-0242ac170004\",\n",
      "        \"library\": \"30e1e0a8-d3e2-11e8-8014-787b8ad942f3\",\n",
      "        \"meta\": {\n",
      "            \"Assay\": \"text mining\",\n",
      "            \"Tissue\": [\n",
      "                {\n",
      "                    \"Name\": \"spinal cord\",\n",
      "                    \"Accession\": \"UBERON:0002240\"\n",
      "                }\n",
      "            ],\n",
      "            \"Readout\": \"mRNA\",\n",
      "            \"Organism\": \"Homo sapiens\",\n",
      "            \"$validator\": \"/@dcic/signature-commons-schema/v4/meta/signature/sigcom.json\",\n",
      "            \"Number_of_genes\": 11398,\n",
      "            \"Original_String\": \"Spinal cord\",\n",
      "            \"Perturbation_Type\": \"shared location\"\n",
      "        }\n",
      "    }\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "## Convert gene names to uuids\n",
    "METADATA_API = \"https://maayanlab.cloud/sigcom/metadata-api\"\n",
    "endpoint = \"/entities/find\"\n",
    "genes = ['UTP14A','S100A6','SCAND1','RRP12','CIAPIN1','ADH5','MTERF3','SPR','CHMP4A','UFM1','VAT1','HACD3','RFC5','COTL1','NPRL2','TRIB3','PCCB','TLE1','CD58','BACE2','KDM3A','TARBP1','RNH1','CHAC1','MBNL2','VDAC1','TES','OXA1L','NOP56','HAT1','CPNE3','DNMT1','ARHGAP1','VPS28','EIF2S2','BAG3','CDCA4','NPDC1','RPS6KA1','FIS1','SYPL1','SARS','CDC45','CANT1','HERPUD1','SORBS3','MRPS2','TOR1A','TNIP1','SLC25A46','MAL','EPCAM','HDAC6','CAPN1','TNRC6B','PKD1','RRS1','HP','ANO10','CEP170B','IDE','DENND2D','CAMK2B','ZNF358','RPP38','MRPL19','NUCB2','GNAI1','LSR','ADGRE2','PKMYT1','CDK5R1','ABL1','PILRB','AXIN1','FBXL8','MCF2L','DBNDD1','IGHMBP2','WIPF2','WFS1','OGFOD2','MAPK1IP1L','COL11A1','REG3A','SERPINA1','MYCBP2','PIGK','TCAP','CRADD','ELK1','DNAJB2','ZBTB16','DAZAP1','MAPKAPK2','EDRF1','CRIP1','UCP3','AGR2','P4HA2']\n",
    "\n",
    "query = {\n",
    "    \"filter\":{\n",
    "        \"where\": {\n",
    "            \"meta.Name\": {\n",
    "                \"inq\": genes\n",
    "            }     \n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "res = requests.post(METADATA_API+ endpoint,  json=query)\n",
    "gene_ids = [i[\"id\"] for i in res.json()]\n",
    "\n",
    "# Signature Search\n",
    "DATA_API = \"https://maayanlab.cloud/sigcom/data-api/api/v1\"\n",
    "endpoint = \"/enrich/overlap\"\n",
    "\n",
    "query = {\n",
    "    \"database\": \"enrichr_geneset\",\n",
    "    \"entities\": gene_ids,\n",
    "    \"limit\": 5\n",
    "}\n",
    "\n",
    "res = requests.post(DATA_API + endpoint, json=query)\n",
    "signatures = [i[\"uuid\"] for i in res.json()[\"results\"]]\n",
    "signatures\n",
    "\n",
    "# Get metadata for signatures\n",
    "\n",
    "endpoint = \"/signatures/find\"\n",
    "query = {\n",
    "    \"filter\":{\n",
    "        \"where\": {\n",
    "            \"id\": {\n",
    "                \"inq\": signatures\n",
    "            }     \n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "res = requests.post(METADATA_API+ endpoint,  json=query)\n",
    "print(json.dumps(res.json(),indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Datatype: rank_matrix\n",
    "**Method** \t`POST`<br>\n",
    "**URL** \t`/enrich/rank`\n",
    "\n",
    "\n",
    "**Query**<br>\n",
    "```\n",
    "{\n",
    "  \"database\": \"string\",\n",
    "  \"entities\": [\n",
    "    \"string\"\n",
    "  ],\n",
    "  \"offset\": 0,\n",
    "  \"limit\": 0\n",
    "}\n",
    "```\n",
    "\n",
    "**Returns** The analysis results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[\n",
      "    {\n",
      "        \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "        \"id\": \"856b137c-9838-454c-860d-7174ccd66b57\",\n",
      "        \"library\": \"1c38c500-3968-474e-b32e-46dc2046087d\",\n",
      "        \"meta\": {\n",
      "            \"id\": \"5141e76cdfe7719b88302457\",\n",
      "            \"Assay\": \"L1000\",\n",
      "            \"sig_id\": \"DOS004_VCAP_24H:BRD-K60056908-001-01-5:5.02\",\n",
      "            \"Readout\": \"gene\",\n",
      "            \"created\": \"2018-11-27T18:17:23.863Z\",\n",
      "            \"is_gold\": true,\n",
      "            \"pert_id\": \"BRD-K60056908\",\n",
      "            \"pool_id\": \"epsilon\",\n",
      "            \"Organism\": \"Homo sapiens\",\n",
      "            \"ds_index\": 120377,\n",
      "            \"zmad_ref\": \"population\",\n",
      "            \"Cell_Line\": [\n",
      "                {\n",
      "                    \"Name\": \"VCaP\",\n",
      "                    \"Accession\": \"CVCL_2235\"\n",
      "                }\n",
      "            ],\n",
      "            \"distil_id\": [\n",
      "                \"DOS004_VCAP_24H_X1_B2_DUO52HI53LO:C04\",\n",
      "                \"DOS004_VCAP_24H_X2_B2_DUO53HI52LO:C04\",\n",
      "                \"DOS004_VCAP_24H_X3_B3_DUO52HI53LO:C04\"\n",
      "            ],\n",
      "            \"distil_ss\": 2.39877,\n",
      "            \"pert_desc\": \"-666\",\n",
      "            \"pert_dose\": 5.02,\n",
      "            \"pert_time\": 24,\n",
      "            \"pert_type\": \"trt_cp\",\n",
      "            \"$validator\": \"/@dcic/signature-commons-schema/v4/meta/signature/clue-io.json\",\n",
      "            \"pert_idose\": \"5 \\u00b5M\",\n",
      "            \"pert_iname\": \"BRD-K60056908\",\n",
      "            \"pert_itime\": \"24 h\",\n",
      "            \"brew_prefix\": \"DOS004_VCAP_24H\",\n",
      "            \"pert_mfc_id\": \"BRD-K60056908-001-01-5\",\n",
      "            \"target_is_lm\": false,\n",
      "            \"distil_cc_q75\": 0.3,\n",
      "            \"distil_nsample\": 3,\n",
      "            \"pert_dose_unit\": \"\\u00b5M\",\n",
      "            \"pert_time_unit\": \"h\",\n",
      "            \"provenance_code\": \"KMV+F2D+INO+QNO+ZSP+MOZ+GBP,D\",\n",
      "            \"Perturbation_Type\": \"small molecules; gene knockdown\",\n",
      "            \"pct_self_rank_q25\": 2.7450980392156863,\n",
      "            \"ngenes_modulated_dn_lm\": 2,\n",
      "            \"ngenes_modulated_up_lm\": 7\n",
      "        }\n",
      "    },\n",
      "    {\n",
      "        \"$validator\": \"/dcic/signature-commons-schema/v5/core/signature.json\",\n",
      "        \"id\": \"753cc339-fa87-4709-885b-fee6447a3b0f\",\n",
      "        \"library\": \"1c38c500-3968-474e-b32e-46dc2046087d\",\n",
      "        \"meta\": {\n",
      "            \"id\": \"5141e385dfe7719b882e94d5\",\n",
      "            \"Assay\": \"L1000\",\n",
      "            \"qc_tag\": [\n",
      "                \"NOT_SELFCON\"\n",
      "            ],\n",
      "            \"sig_id\": \"CPC005_MCF7_6H:BRD-A54445976-001-02-5:10\",\n",
      "            \"Readout\": \"gene\",\n",
      "            \"created\": \"2018-11-27T18:02:13.222Z\",\n",
      "            \"is_gold\": false,\n",
      "            \"pert_id\": \"BRD-A54445976\",\n",
      "            \"pool_id\": \"epsilon\",\n",
      "            \"Organism\": \"Homo sapiens\",\n",
      "            \"ds_index\": 26190,\n",
      "            \"zmad_ref\": \"population\",\n",
      "            \"Cell_Line\": [\n",
      "                {\n",
      "                    \"Name\": \"MCF-7\",\n",
      "                    \"Accession\": \"CVCL_0031\"\n",
      "                }\n",
      "            ],\n",
      "            \"distil_id\": [\n",
      "                \"CPC005_MCF7_6H_X1_B3_DUO52HI53LO:I21\",\n",
      "                \"CPC005_MCF7_6H_X2_B3_DUO52HI53LO:I21\",\n",
      "                \"CPC005_MCF7_6H_X2_F1B3_DUO52HI53LO:I21\",\n",
      "                \"CPC005_MCF7_6H_X3_B3_DUO52HI53LO:I21\",\n",
      "                \"CPC005_MCF7_6H_X5_F1B5_DUO52HI53LO:I21\"\n",
      "            ],\n",
      "            \"distil_ss\": 3.64756,\n",
      "            \"pert_desc\": \"Napelline\",\n",
      "            \"pert_dose\": 10,\n",
      "            \"pert_time\": 6,\n",
      "            \"pert_type\": \"trt_cp\",\n",
      "            \"$validator\": \"/@dcic/signature-commons-schema/v4/meta/signature/clue-io.json\",\n",
      "            \"pert_idose\": \"10 \\u00b5M\",\n",
      "            \"pert_iname\": \"napelline\",\n",
      "            \"pert_itime\": \"6 h\",\n",
      "            \"brew_prefix\": \"CPC005_MCF7_6H\",\n",
      "            \"pert_mfc_id\": \"BRD-A54445976-001-02-5\",\n",
      "            \"target_is_lm\": false,\n",
      "            \"distil_cc_q75\": 0.15,\n",
      "            \"distil_nsample\": 5,\n",
      "            \"pert_dose_unit\": \"\\u00b5M\",\n",
      "            \"pert_time_unit\": \"h\",\n",
      "            \"provenance_code\": \"KMV+F2D+INO+QNO+ZSP+MOZ+GBP,D\",\n",
      "            \"Perturbation_Type\": \"small molecules; gene knockdown\",\n",
      "            \"pct_self_rank_q25\": 4.62962962962963,\n",
      "            \"ngenes_modulated_dn_lm\": 27,\n",
      "            \"ngenes_modulated_up_lm\": 20\n",
      "        }\n",
      "    }\n",
      "]\n"
     ]
    }
   ],
   "source": [
    "## Convert gene names to uuids\n",
    "METADATA_API = \"https://maayanlab.cloud/sigcom/metadata-api\"\n",
    "endpoint = \"/entities/find\"\n",
    "genes = ['UTP14A','S100A6','SCAND1','RRP12','CIAPIN1','ADH5','MTERF3','SPR','CHMP4A','UFM1','VAT1','HACD3','RFC5','COTL1','NPRL2','TRIB3','PCCB','TLE1','CD58','BACE2','KDM3A','TARBP1','RNH1','CHAC1','MBNL2','VDAC1','TES','OXA1L','NOP56','HAT1','CPNE3','DNMT1','ARHGAP1','VPS28','EIF2S2','BAG3','CDCA4','NPDC1','RPS6KA1','FIS1','SYPL1','SARS','CDC45','CANT1','HERPUD1','SORBS3','MRPS2','TOR1A','TNIP1','SLC25A46','MAL','EPCAM','HDAC6','CAPN1','TNRC6B','PKD1','RRS1','HP','ANO10','CEP170B','IDE','DENND2D','CAMK2B','ZNF358','RPP38','MRPL19','NUCB2','GNAI1','LSR','ADGRE2','PKMYT1','CDK5R1','ABL1','PILRB','AXIN1','FBXL8','MCF2L','DBNDD1','IGHMBP2','WIPF2','WFS1','OGFOD2','MAPK1IP1L','COL11A1','REG3A','SERPINA1','MYCBP2','PIGK','TCAP','CRADD','ELK1','DNAJB2','ZBTB16','DAZAP1','MAPKAPK2','EDRF1','CRIP1','UCP3','AGR2','P4HA2']\n",
    "\n",
    "query = {\n",
    "    \"filter\":{\n",
    "        \"where\": {\n",
    "            \"meta.Name\": {\n",
    "                \"inq\": genes\n",
    "            }     \n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "res = requests.post(METADATA_API+ endpoint,  json=query)\n",
    "gene_ids = [i[\"id\"] for i in res.json()]\n",
    "\n",
    "# Signature Search\n",
    "DATA_API = \"https://maayanlab.cloud/sigcom/data-api/api/v1\"\n",
    "endpoint = \"/enrich/rank\"\n",
    "\n",
    "query = {\n",
    "    \"database\": \"lincs_clue\",\n",
    "    \"entities\": gene_ids,\n",
    "    \"limit\": 2\n",
    "}\n",
    "\n",
    "res = requests.post(DATA_API + endpoint, json=query)\n",
    "signatures = [i[\"uuid\"] for i in res.json()[\"results\"]]\n",
    "signatures\n",
    "\n",
    "# Get metadata for signatures\n",
    "\n",
    "endpoint = \"/signatures/find\"\n",
    "query = {\n",
    "    \"filter\":{\n",
    "        \"where\": {\n",
    "            \"id\": {\n",
    "                \"inq\": signatures\n",
    "            }     \n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "res = requests.post(METADATA_API+ endpoint,  json=query)\n",
    "print(json.dumps(res.json(),indent=4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Up and Down Gene Set\n",
    "**Method** \t`POST`<br>\n",
    "**URL** \t`/enrich/ranktwosided`\n",
    "\n",
    "\n",
    "**Query**<br>\n",
    "```\n",
    "{\n",
    "  \"database\": \"string\",\n",
    "  \"up_entities\": [\n",
    "    \"string\"\n",
    "  ],\n",
    "  \"down_entities\": [\n",
    "    \"string\"\n",
    "  ],\n",
    "  \"offset\": 0,\n",
    "  \"limit\": 0\n",
    "}\n",
    "```\n",
    "\n",
    "**Returns** The analysis results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "JSONDecodeError",
     "evalue": "Expecting value: line 1 column 1 (char 0)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mJSONDecodeError\u001b[0m                           Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-11-aba225eb188b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     33\u001b[0m \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrequests\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpost\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mDATA_API\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mendpoint\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mquery\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 34\u001b[0;31m \u001b[0msignatures\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"uuid\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjson\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"results\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     35\u001b[0m \u001b[0msignatures\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     36\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Erol/jupyter-template/jupytervenv/lib/python3.8/site-packages/requests/models.py\u001b[0m in \u001b[0;36mjson\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m    896\u001b[0m                     \u001b[0;31m# used.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    897\u001b[0m                     \u001b[0;32mpass\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 898\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mcomplexjson\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloads\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtext\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    899\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    900\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/__init__.py\u001b[0m in \u001b[0;36mloads\u001b[0;34m(s, cls, object_hook, parse_float, parse_int, parse_constant, object_pairs_hook, **kw)\u001b[0m\n\u001b[1;32m    355\u001b[0m             \u001b[0mparse_int\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mparse_float\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m \u001b[0;32mand\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    356\u001b[0m             parse_constant is None and object_pairs_hook is None and not kw):\n\u001b[0;32m--> 357\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0m_default_decoder\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdecode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    358\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mcls\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    359\u001b[0m         \u001b[0mcls\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mJSONDecoder\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/decoder.py\u001b[0m in \u001b[0;36mdecode\u001b[0;34m(self, s, _w)\u001b[0m\n\u001b[1;32m    335\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    336\u001b[0m         \"\"\"\n\u001b[0;32m--> 337\u001b[0;31m         \u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraw_decode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0midx\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0m_w\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    338\u001b[0m         \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_w\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    339\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mend\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/decoder.py\u001b[0m in \u001b[0;36mraw_decode\u001b[0;34m(self, s, idx)\u001b[0m\n\u001b[1;32m    353\u001b[0m             \u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mscan_once\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0midx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    354\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 355\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mJSONDecodeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Expecting value\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0ms\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    356\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mJSONDecodeError\u001b[0m: Expecting value: line 1 column 1 (char 0)"
     ]
    }
   ],
   "source": [
    "## Convert gene names to uuids\n",
    "METADATA_API = \"https://maayanlab.cloud/sigcom/metadata-api\"\n",
    "endpoint = \"/entities/find\"\n",
    "up_genes = ['UTP14A','S100A6','SCAND1','RRP12','CIAPIN1','ADH5','MTERF3','SPR','CHMP4A','UFM1','VAT1','HACD3','RFC5','COTL1','NPRL2','TRIB3','PCCB','TLE1','CD58','BACE2','KDM3A','TARBP1','RNH1','CHAC1','MBNL2','VDAC1','TES','OXA1L','NOP56','HAT1','CPNE3','DNMT1','ARHGAP1','VPS28','EIF2S2','BAG3','CDCA4','NPDC1','RPS6KA1','FIS1','SYPL1','SARS','CDC45','CANT1','HERPUD1','SORBS3','MRPS2','TOR1A','TNIP1','SLC25A46']\n",
    "down_genes = ['MAL','EPCAM','HDAC6','CAPN1','TNRC6B','PKD1','RRS1','HP','ANO10','CEP170B','IDE','DENND2D','CAMK2B','ZNF358','RPP38','MRPL19','NUCB2','GNAI1','LSR','ADGRE2','PKMYT1','CDK5R1','ABL1','PILRB','AXIN1','FBXL8','MCF2L','DBNDD1','IGHMBP2','WIPF2','WFS1','OGFOD2','MAPK1IP1L','COL11A1','REG3A','SERPINA1','MYCBP2','PIGK','TCAP','CRADD','ELK1','DNAJB2','ZBTB16','DAZAP1','MAPKAPK2','EDRF1','CRIP1','UCP3','AGR2','P4HA2']\n",
    "\n",
    "query = {\n",
    "    \"filter\":{\n",
    "        \"where\": {\n",
    "            \"meta.Name\": {\n",
    "                \"inq\": up_genes + down_genes\n",
    "            }     \n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "res = requests.post(METADATA_API+ endpoint,  json=query)\n",
    "results = res.json()\n",
    "up_gene_ids = [i[\"id\"] for i in results if i[\"meta\"][\"Name\"] in up_genes]\n",
    "down_gene_ids = [i[\"id\"] for i in results if i[\"meta\"][\"Name\"] in down_genes]\n",
    "\n",
    "# Signature Search\n",
    "DATA_API = \"https://maayanlab.cloud/sigcom/data-api/api/v1\"\n",
    "endpoint = \"/enrich/ranktwosided\"\n",
    "\n",
    "query = {\n",
    "    \"database\": \"lincs_clue\",\n",
    "    \"up_entities\": up_gene_ids,\n",
    "    \"down_entities\": down_gene_ids,\n",
    "    \"limit\": 2\n",
    "}\n",
    "\n",
    "res = requests.post(DATA_API + endpoint, json=query)\n",
    "signatures = [i[\"uuid\"] for i in res.json()[\"results\"]]\n",
    "signatures\n",
    "\n",
    "# Get metadata for signatures\n",
    "\n",
    "endpoint = \"/signatures/find\"\n",
    "query = {\n",
    "    \"filter\":{\n",
    "        \"where\": {\n",
    "            \"id\": {\n",
    "                \"inq\": signatures\n",
    "            }     \n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "res = requests.post(METADATA_API+ endpoint,  json=query)\n",
    "print(json.dumps(res.json(),indent=4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter-default",
   "language": "python",
   "name": "jupyter-default"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
